File: backend/app/password_policy.py
PASSWORD_POLICY_MESSAGE = (
    "Password must be at least 10 characters and include at least 3 of: "
    "uppercase letter, lowercase letter, number, symbol"
)


def password_policy_status(password: str) -> dict[str, bool]:
    value = str(password or "")
    return {
        "length": len(value) >= 10,
        "uppercase": any(character.isupper() for character in value),
        "lowercase": any(character.islower() for character in value),
        "number": any(character.isdigit() for character in value),
        "symbol": any(not character.isalnum() and not character.isspace() for character in value),
    }


def password_meets_policy(password: str) -> bool:
    status = password_policy_status(password)
    category_count = sum(
        1
        for key in ("uppercase", "lowercase", "number", "symbol")
        if status[key]
    )
    return status["length"] and category_count >= 3

File: frontend/src/features/auth/passwordPolicy.js
export const passwordPolicyMessage =
  "Use at least 10 characters and include at least 3 of: uppercase letter, lowercase letter, number, symbol.";

export function getPasswordPolicyStatus(password) {
  const value = String(password || "");
  return [
    { group: "required", label: "At least 10 characters", met: value.length >= 10 },
    { group: "category", label: "Uppercase letter", met: /[A-Z]/.test(value) },
    { group: "category", label: "Lowercase letter", met: /[a-z]/.test(value) },
    { group: "category", label: "Number", met: /\d/.test(value) },
    { group: "category", label: "Symbol", met: /[^\sA-Za-z0-9]/.test(value) }
  ];
}

export function validatePassword(password) {
  const status = getPasswordPolicyStatus(password);
  const categoryCount = status.slice(1).filter((item) => item.met).length;
  return {
    isValid: status[0].met && categoryCount >= 3,
    message: passwordPolicyMessage,
    status
  };
}
